prusti-interface/src/specs/checker.rs

```diff
@@ -0,0 +1,118 @@
+use log::{info, debug};
+use rustc_middle::{
+    hir::map::Map,
+    ty::TyCtxt,
+};
+use rustc_span::{Span, MultiSpan};
+use rustc_hir::{
+    self as hir,
+    intravisit::{self, Visitor},
+    itemlikevisit::ItemLikeVisitor,
+    def_id::{DefId, LocalDefId},
+};
+
+use std::collections::HashMap;
+
+use crate::{
+    environment::Environment,
+    utils::has_prusti_attr,
+    PrustiError,
+};
+
+
+pub struct SpecChecker<'tcx> {
+    tcx: TyCtxt<'tcx>,
+
+    predicates: HashMap<DefId, Span>,
+    // span of call and definition of predicates illegally called
+    pred_calls: Vec<(Span, Span)>,
+}
+
+// shallow visitor, just to collect all function items that originate from
+// predicates
+impl<'tcx> ItemLikeVisitor<'tcx> for SpecChecker<'tcx> {
+    fn visit_item(&mut self, item: &'tcx hir::Item<'tcx>) {
+        // collect DefIds for all predicate function items
+        if has_prusti_attr(item.attrs, "pred_spec_id_ref") {
+            let def_id = self.tcx.hir().local_def_id(item.hir_id).to_def_id();
+            self.predicates.insert(def_id, item.span);
+        }
+    }
+
+    fn visit_trait_item(&mut self, _: &'tcx hir::TraitItem<'tcx>) {
+        // nothing here
```

Comment:
do we need some default action here to visit sub-items? should also add a test for it.

Reply:
seems the ItemLikeVisitor is too shallow here, only visiting toplevel items; rewriting it using `intravisit::Visitor`.